### ISSUE TITLE
salt-ssh file.recurse adds a lot of unwanted directories

### ISSUE THREAD

**Issue by anlutro** [saltstack/salt#40594]
I have this state:

```
/var/www/default:
  file.recurse:
    - source: salt://web/default
```

This used to work as you'd expect, but recently I updated salt on my local system and something is horribly broken:

```
          ID: /var/www/default
    Function: file.recurse
      Result: True
     Comment: Recursively updated /var/www/default
     Started: 14:55:09.901724
    Duration: 92.777 ms
     Changes:   
              ----------
              /var/www/default/alt-ssh/odin/files/base/web/default:
                  ----------
                  /var/www/default/alt-ssh/odin/files/base/web/default:
                      New Dir
              /var/www/default/alt-ssh/odin/files/base/web/default/index.html:
                  ----------
                  diff:
                      New file
                  mode:
                      0644
              /var/www/default/alt-ssh/odin/files/base/web/default/robots.txt:
                  ----------
                  diff:
                      New file
                  mode:
                      0644
```

The git pull I did shows `Updating bb37f133fc..4fa58be222` so I guess the error is in there somewhere. If I check out bb37f133fc everything works fine.

```
$ salt-ssh --versions
Salt Version:
           Salt: 2016.11.3-158-g4fa58be
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Jan 19 2017, 14:48:08)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.1.6
 
System Versions:
           dist: debian 9.0 
        machine: x86_64
        release: 4.9.0-2-amd64
         system: Linux
        version: debian 9.0
```

**Comment by anlutro** [comment]
`git bisect` is telling me that 8367735063eaa12bd42ae71b8922119ce747f4c6 is the commit that's causing this bug.

**Comment by anlutro** [comment]
Indeed, if I revert that commit, everything seems to work as before.

**Comment by gtmanfred** [comment]
If we revert that commit, then we get the salt master moving all the fileserver cache back to /var/tmp/*_salt/running_data  instead of being in /var/cache/salt/master, so we need to come up with another solution.

Let me do some testing later today.

**Comment by Ch3LL** [comment]
This has been resolved with #40642. Closing
